### PR TITLE
Type the query builder

### PR DESCRIPTION
**Pull request by ChristophWurst**
This is mostly moving types from just phpdoc to actual php type hints.

But I've also changed update, delete and insert slightly so the first parameter is always a string and not nullable. I really can't see a use case where a null table name would be ok.

Changes

* Added return type hints
* Added parameter type hints
* Fixed injected `Connection` because we can't pass our implementation to `\Doctrine\DBAL\Query\QueryBuilder::__construct`, though it indirectly implements that class
* Enabled strict types for both files

### CODE REVIEW

**Review by rullzer — APPROVED** [review]

File: lib/public/DB/QueryBuilder/IQueryBuilder.php [hunk 1]
```diff
@@ -206,15 +211,15 @@ public function setParameter($key, $value, $type = null);
...
+	public function setParameters(array $params, array $types = []) :self;
 
 	/**
 	 * Gets all defined query parameters for the query being constructed indexed by parameter index or name.
 	 *
 	 * @return array The currently defined query parameters indexed by parameter index or name.
 	 * @since 8.2.0
 	 */
-	public function getParameters();
+	public function getParameters(): array ;
```
Comment (nickvergessen) [inline lib/public/DB/QueryBuilder/IQueryBuilder.php:222]:
```suggestion
	public function getParameters(): array;
```

File: lib/public/DB/QueryBuilder/IQueryBuilder.php [hunk 2]
```diff
@@ -232,7 +237,7 @@ public function getParameter($key);
 	 * @return array The currently defined query parameter types indexed by parameter index or name.
 	 * @since 8.2.0
 	 */
-	public function getParameterTypes();
+	public function getParameterTypes(): array ;
```
Comment (nickvergessen) [inline lib/public/DB/QueryBuilder/IQueryBuilder.php:240]:
```suggestion
	public function getParameterTypes(): array;
```

File: lib/public/DB/QueryBuilder/IQueryBuilder.php [hunk 3]
```diff
@@ -443,12 +448,12 @@ public function from($from, $alias = null);
...
 	 * @param string $join The table name to join.
 	 * @param string $alias The alias of the join table.
-	 * @param string $condition The condition for the join.
+	 * @param string|IStringSerializable $condition The condition for the join.
 	 *
 	 * @return $this This QueryBuilder instance.
 	 * @since 8.2.0
 	 */
-	public function join($fromAlias, $join, $alias, $condition = null);
+	public function join(string $fromAlias, string $join, string $alias, $condition): self;
```
Comment (nickvergessen) [inline lib/public/DB/QueryBuilder/IQueryBuilder.php:456]: This breaks if no condition was specified.
Now I know it makes little sense, but it is an incompatibility with the doctrine thing, so not sure.

File: lib/public/DB/QueryBuilder/IQueryBuilder.php [hunk 4]
```diff
@@ -743,12 +748,12 @@ public function addOrderBy($sort, $order = null);
...
 	 * @since 8.2.0
 	 */
-	public function getQueryPart($queryPartName);
+	public function getQueryPart(string $queryPartName);
 
 	/**
 	 * Gets all query parts.
 	 *
-	 * @return array
+	 * @return mixed[]
```
Comment (nickvergessen) [inline lib/public/DB/QueryBuilder/IQueryBuilder.php:756]:
```suggestion
	 * @return string[]
```

File: lib/public/DB/QueryBuilder/IQueryBuilder.php [hunk 5]
```diff
@@ -873,15 +878,15 @@ public function createParameter($name);
 	 * @return IQueryFunction
 	 * @since 8.2.0
 	 */
-	public function createFunction($call);
+	public function createFunction($call):IQueryFunction;
```
Comment (nickvergessen) [inline lib/public/DB/QueryBuilder/IQueryBuilder.php:881]:
```suggestion
	public function createFunction($call): IQueryFunction;
```

File: lib/private/DB/QueryBuilder/QueryBuilder.php [hunk 6]
```diff
@@ -433,7 +440,7 @@ public function selectAlias($select, $alias) {
 	 *
 	 * @return $this This QueryBuilder instance.
 	 */
-	public function selectDistinct($select) {
+	public function selectDistinct($select) : IQueryBuilder {
```
Comment (nickvergessen) [inline lib/private/DB/QueryBuilder/QueryBuilder.php:443]:
```suggestion
	public function selectDistinct($select): IQueryBuilder {
```

File: lib/private/DB/QueryBuilder/QueryBuilder.php [hunk 7]
```diff
@@ -456,7 +463,7 @@ public function selectDistinct($select) {
 	 *
 	 * @return $this This QueryBuilder instance.
 	 */
-	public function addSelect(...$selects) {
+	public function addSelect(...$selects) : IQueryBuilder {
```
Comment (nickvergessen) [inline lib/private/DB/QueryBuilder/QueryBuilder.php:466]:
```suggestion
	public function addSelect(...$selects): IQueryBuilder {
```

File: lib/private/DB/QueryBuilder/QueryBuilder.php [hunk 8]
```diff
@@ -480,11 +487,11 @@ public function addSelect(...$selects) {
 	 * </code>
 	 *
 	 * @param string $delete The table whose rows are subject to the deletion.
-	 * @param string $alias The table alias used in the constructed query.
+	 * @param string|null $alias The table alias used in the constructed query.
 	 *
 	 * @return $this This QueryBuilder instance.
 	 */
-	public function delete($delete = null, $alias = null) {
+	public function delete(string $delete, string $alias = null) : IQueryBuilder {
```
Comment (nickvergessen) [inline lib/private/DB/QueryBuilder/QueryBuilder.php:494]:
```suggestion
	public function delete(string $delete, string $alias = null): IQueryBuilder {
```

File: lib/private/DB/QueryBuilder/QueryBuilder.php [hunk 9]
```diff
@@ -505,11 +512,11 @@ public function delete($delete = null, $alias = null) {
 	 * </code>
 	 *
 	 * @param string $update The table whose rows are subject to the update.
-	 * @param string $alias The table alias used in the constructed query.
+	 * @param string|string $alias The table alias used in the constructed query.
 	 *
 	 * @return $this This QueryBuilder instance.
 	 */
-	public function update($update = null, $alias = null) {
+	public function update(string $update, string $alias = null) : IQueryBuilder {
```
Comment (nickvergessen) [inline lib/private/DB/QueryBuilder/QueryBuilder.php:519]:
```suggestion
	public function update(string $update, string $alias = null): IQueryBuilder {
```

File: lib/private/DB/QueryBuilder/QueryBuilder.php [hunk 10]
```diff
@@ -694,11 +701,11 @@ public function rightJoin($fromAlias, $join, $alias, $condition = null) {
 	 * </code>
 	 *
 	 * @param string $key The column to set.
-	 * @param string $value The value, expression, placeholder, etc.
+	 * @param string|IStringSerializable $value The value, expression, placeholder, etc.
 	 *
 	 * @return $this This QueryBuilder instance.
 	 */
-	public function set($key, $value) {
+	public function set(string $key, $value) : IQueryBuilder {
```
Comment (nickvergessen) [inline lib/private/DB/QueryBuilder/QueryBuilder.php:708]:
```suggestion
	public function set(string $key, $value): IQueryBuilder {
```

File: lib/private/DB/QueryBuilder/QueryBuilder.php [hunk 11]
```diff
@@ -760,7 +767,7 @@ public function where(...$predicates) {
 	 *
 	 * @see where()
 	 */
-	public function andWhere(...$where) {
+	public function andWhere(...$where) : IQueryBuilder {
```
Comment (nickvergessen) [inline lib/private/DB/QueryBuilder/QueryBuilder.php:770]:
```suggestion
	public function andWhere(...$where): IQueryBuilder {
```

File: lib/private/DB/QueryBuilder/QueryBuilder.php [hunk 12]
```diff
@@ -867,14 +874,14 @@ public function addGroupBy(...$groupBys) {
 	 * </code>
 	 *
 	 * @param string $column The column into which the value should be inserted.
-	 * @param string $value The value that should be inserted into the column.
+	 * @param string|IParameter $value The value that should be inserted into the column.
 	 *
 	 * @return $this This QueryBuilder instance.
 	 */
-	public function setValue($column, $value) {
+	public function setValue(string $column, $value) : IQueryBuilder {
```
Comment (nickvergessen) [inline lib/private/DB/QueryBuilder/QueryBuilder.php:881]:
```suggestion
	public function setValue(string $column, $value): IQueryBuilder {
```

File: lib/private/DB/QueryBuilder/QueryBuilder.php [hunk 13]
```diff
@@ -952,7 +959,7 @@ public function andHaving(...$having) {
 	 *
 	 * @return $this This QueryBuilder instance.
 	 */
-	public function orHaving(...$having) {
+	public function orHaving(...$having) : IQueryBuilder {
```
Comment (nickvergessen) [inline lib/private/DB/QueryBuilder/QueryBuilder.php:962]: There are some more ` :  ` issues, does php cs fix them?

**Review by nickvergessen — CHANGES_REQUESTED** [review]

File: lib/public/DB/QueryBuilder/IQueryBuilder.php (hunk 3 above)
Comment (ChristophWurst) [reply lib/public/DB/QueryBuilder/IQueryBuilder.php:456]: Yeah, I was not fully sure about this one. BUT we can enforce having a condition. We'll still be compatible with doctrine. And I really don't see any use case except the cross product where a join without condition would work. Do you?

File: lib/private/DB/QueryBuilder/QueryBuilder.php (hunk 13 above)
Comment (ChristophWurst) [reply lib/private/DB/QueryBuilder/QueryBuilder.php:962]: I think

**Comment by MorrisJobke** [comment]
@ChristophWurst 🏓

**Comment by ChristophWurst** [comment]
I think CI is really not happy about this. I have to split this into smaller patches, it's too complex to do in one go.

**Pull request closed without merging**